### PR TITLE
add Equality.jl with method for edges and added tests for MinimumEdgeDominatingSet

### PR DESCRIPTION
**Pull request by joshyaffee**
I addded a new file for methods that check equlity. This is because the types of some of our objects are unions of types, so these methods can help compare between these representations. The only method so far is for undurected edge comparison.

This method is used in unit tests... more to come soon! I also think I have a more valid way of testing the domination methods, kind of by inspection and case by case, but it doesn't use the logic of the methods. By inspection is still a bit suspect... what else should you do for unit tests, though?

# Pull Request Checklist

Thank you for contributing to `GraphProperties.jl`!

## Documentation and Style

- [X] I have added docstrings to any new functions or types, as applicable.
- [X] My code follows the project's coding style and guidelines.

## New Features and Code Changes

- [X] If you have added a new algorithm or function, you have also included corresponding unit tests.
- [X] I confirm that any code added is my own work and is not copied from other sources.

## Review Process

- [X] I have requested a review from at least one other contributor.

## Additional Notes

- Please describe any additional changes or information relevant to this PR.

---

### Developer's Declaration:

By submitting this Pull Request, I confirm the following:

- [X] The code I am submitting is my own work and I have the rights to use and distribute it.
- [X] I confirm that the code I have added is not copied from other sources and that I have appropriately cited or acknowledged any external code or references used.
- [X] I understand that my code will be reviewed and any changes requested must be addressed.

## PR Type

What kind of change does this PR introduce?
- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation content changes
- [X] Unit Testing
- [ ] Other... Please describe:

## What is the new behavior (if this is a feature change)?

- [X] Yes: added first equals method
- [ ] No

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

File: src/Invariants/Equality.jl
Comment (RandyRDavila) [inline src/Invariants/Equality.jl:26]: A more Julia way of doing this would be with the following:

```julia 
length(a) != length(b) && return false 
```

File: src/Invariants/Equality.jl
Comment (RandyRDavila) [inline src/Invariants/Equality.jl:32]: Here again:

```julia 
e ∉ b && return false 
```

File: src/Invariants/Equality.jl
Comment (RandyRDavila) [inline src/Invariants/Equality.jl:11]: Complete sentence:

```julia 
# Convert `a` to an array of tuples. 
```

File: src/Invariants/Equality.jl
Comment (RandyRDavila) [inline src/Invariants/Equality.jl:36]: Remove this empty line.

File: src/Invariants/Equality.jl
Comment (RandyRDavila) [reply src/Invariants/Equality.jl:32]: BTW `∉` can be made by typing \notin + TAB

File: src/Invariants/Equality.jl
Comment (joshyaffee) [reply src/Invariants/Equality.jl:26]: Is this equivalent to
`length(a) == length(b) || return false`
and if so is there a preference?

File: src/Invariants/Equality.jl
Comment (joshyaffee) [reply src/Invariants/Equality.jl:32]: Same question as above

File: src/Invariants/Equality.jl
Comment (RandyRDavila) [inline src/Invariants/Equality.jl:2]: Maybe try:

```julia
const SimpleInt64Edge = Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}
```

Then see my comments below

File: src/Invariants/Equality.jl
Comment (RandyRDavila) [inline src/Invariants/Equality.jl:19]: Per my comment at the top (the one defining `SimpleInt64Edge`):

```julia 
isa(b, SimpleInt64Edge)  && b = [(min(e.src, e.dst), max(e.src, e.dst)) for e in b]
!isa(b, SimpleInt64Edge) && b = [(min(e[1], e[2]), max(e[1], e[2])) for e in b]
```

File: src/Invariants/Equality.jl
Comment (RandyRDavila) [reply src/Invariants/Equality.jl:32]: Why not replace checking the `length` and the element wise containment with :

```julia
a != b && return false 
return true 
```

**Review by RandyRDavila — CHANGES_REQUESTED** [review]
Please see my comments. Also, unless `Equality.jl` is supposed to be a Julia module, we should use lower case name - `equality.jl`. After you address my comments I'll approve your PR. Thanks so much for this!

File: src/Invariants/Equality.jl
Comment (RandyRDavila) [reply src/Invariants/Equality.jl:26]: Use DeMorgans law for an equivalence

File: src/Invariants/Equality.jl
Comment (joshyaffee) [reply src/Invariants/Equality.jl:32]: Since order does not matter. I think I will change it to use sets so we do not have to deal with this.

File: src/Invariants/Equality.jl
Comment (RandyRDavila) [reply src/Invariants/Equality.jl:26]: DeMorgan's Law gives: 

```julia 
length(a) == length(b) || return true
```

as an equivalence

Comment (RandyRDavila) [reply src/Invariants/Equality.jl:26]: Also, there is no preference. I just typically use `&&` when working in julia

Comment (joshyaffee) [reply src/Invariants/Equality.jl:26]: I'm not sure I understand how to apply DeMorgans law to short circuit logic. This is the way I see it:
We want to short circuit when the lengths are the same and to return false otherwise.

so in one case `length(a) != length(b) && return false` short circuits when `length(a) != length(b)` is not true or when the lengths are the same. Otherwise, return false.

In the case I offered, `length(a) == length(b) || return false` short circuits if the first condition is true; that is once again when the lengths are the same. Otherwise, return false.

Have I made a mistake?

File: src/Invariants/Equality.jl
Comment (RandyRDavila) [reply src/Invariants/Equality.jl:19]: `isa` is a really useful julia function for checking types. I actually prefer 

```julia 
b isa SimpleInt64Edge
```

over 

```julia 
isa(b, SimpleInt64Edge)
```

But because we had a negation, I decided on the latter

File: src/Invariants/Equality.jl
Comment (joshyaffee) [reply src/Invariants/Equality.jl:26]: Seeing your example, I think I see what you mean: `length(a) == length(b) || return true` is the negation of the original statement, so we can flip the return back to false to get a statement of equivalent function?

Comment (joshyaffee) [reply src/Invariants/Equality.jl:26]: Also I am starting to see why `&&` is more natural even if the first statement is a negative. It reads better as the if-then control flow we are aiming for

File: src/Invariants/Equality.jl
Comment (joshyaffee) [reply src/Invariants/Equality.jl:32]: Is the suggestion above more in line with Julia's style than simply `return a==b` ?

**Comment by joshyaffee** [comment]
New commit is pushed; I think all issues have been addressed. I do have a question, though @RandyRDavila . Would it benefit compilation to declare `a_set` and `b_set` at the top of the function since they get initialized using short circuit logic?

Also, I see the automatic checks have failed, but I don't understnad why... I'll try to look into it.

File: src/Invariants/Equality.jl
Comment (RandyRDavila) [inline src/Invariants/Equality.jl:1]: I like what you've done! However, once I started thinking about it, I realized what you really wanted to do was to _override_ `==`. This can be done because every operation in Julia is a function. So we write new methods for `==`. For example:

```julia 
import Base: ==
using Graphs.SimpleGraphs: SimpleEdge

# Define a helper function to normalize an edge. Notice that
# I prefix the function name with an underscore to indicate
# that it is a helper function and is not meant to be used
# outside of this file.
function _normalize_edge(edge)
    edge isa Tuple && return (min(edge...), max(edge...))
    return (min(edge.src, edge.dst), max(edge.src, edge.dst))
end

function ==(
    a::Union{Vector{Tuple{Int, Int}}, Vector{SimpleEdge{Int64}}},
    b::Union{Vector{Tuple{Int, Int}}, Vector{SimpleEdge{Int64}}},
)
    a_set = Set(_normalize_edge(edge) for edge in a)
    b_set = Set(_normalize_edge(edge) for edge in b)

    return a_set == b_set
end
```

Notice how we write a new method for `==` which acts on the objects we are concerned with.  Moreover, we will no longer need to export `equals` or in this case `==` from `GraphProperties.jl` since overloading base functions like == doesn't require exporting them, because they are already part of the Base module which is universally accessible. When you overload == for specific types, Julia will use your custom definition automatically when those types are compared using ==.

So, if you override == for particular types in your GraphProperties.jl script, you don't need to export ==. As long as the types for which you're overloading == are accessible (i.e., either defined in Base, imported from another module, or exported from your module), your overloaded method will be used.

Sorry for drawing out this PR, but I think doing this early on will make you a much better programmer in the long run. 

Also, I asked GPT4 to help write a good doc string (which I had to edit a little bit). The Julia example for the new method you can literally copy past into a REPL and check that it works. However, its not going to render well in GitHub and might look funny. Hopefully you can see what it should be 

```julia
"""
    ==(
        a::Union{Vector{Tuple{Int, Int}}, Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}},
        b::Union{Vector{Tuple{Int, Int}}, Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}},
    )::Bool

Determine if two sets of edges are equivalent in an undirected graph context.

This method treats each edge as undirected, meaning the order of vertices in
each edge tuple or `Graphs.SimpleGraphs.SimpleEdge` is ignored. Two sets of edges
are considered equal if they contain the same pairs of vertices, regardless of
the order in which the vertices appear in each edge.

# Arguments
- `a`: A vector of edges, either as tuples or `Graphs.SimpleGraphs.SimpleEdge{Int64}`.
- `b`: Another vector of edges to compare against `a`.

# Returns
- `true` if both sets of edges represent the same undirected connections, `false` otherwise.

# Examples
```julia
julia> using Graphs.SimpleGraphs: SimpleEdge

julia> using GraphProperties.Invariants

julia> edges_a = [(1, 2), (3, 4)]
2-element Vector{Tuple{Int64, Int64}}:
 (1, 2)
 (3, 4)

julia> edges_b = [SimpleEdge(2, 1), SimpleEdge(4, 3)]
2-element Vector{SimpleEdge{Int64}}:
 Edge 2 => 1
 Edge 4 => 3

julia> edges_a == edges_b
true
```
"""
```

**Review by RandyRDavila — CHANGES_REQUESTED** [review]
Please see my latest comment. Good work on adding this functionality. It was needed. Also, and again, sorry for drawing out this PR. I really just want to make sure I give you the best feedback I can early on to make you a better programmer :)

**Comment by RandyRDavila** [comment]
@joshyaffee If the Tests ever fail but they are passing locally, then just restart the tests by clicking on the details button and then rerun failed jobs.

File: src/Invariants/Equality.jl
Comment (joshyaffee) [reply src/Invariants/Equality.jl:1]: Would this override `==` for any 

```
a::Union{Vector{Tuple{Int, Int}}, Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}},
b::Union{Vector{Tuple{Int, Int}}, Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}}
```
?

I'm curious if this is not what we want since we may want to compare other vectors of edges where order matters. For example, if we want to compare two paths, we would probably care about order.

Maybe it's better to override `==` for `AbstractOptimalEdgeSet` instead but that leads to the issue that we wouldn't be able to construct instances for unit tests (maybe, let's make constructors for the concrete subtypes?)

File: src/Invariants/Equality.jl
Comment (RandyRDavila) [reply src/Invariants/Equality.jl]: haha silly me..... it should be `return a == b`

File: src/Invariants/Equality.jl
Comment (RandyRDavila) [reply src/Invariants/Equality.jl:1]: Yes, this overrides `==` for any 

```julia
a::Union{Vector{Tuple{Int, Int}}, Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}},
b::Union{Vector{Tuple{Int, Int}}, Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}}
```

However, I think we will be ok... We could always define custom types to represent what we want. For example, 

```julia
abstract type AbstractGraphPath{T <: Int} end

struct GraphPath{T <: Int} <: AbstractGraphPath{T}
    vertices::Vector{T}
    edges::Vector{Tuple{T, T}}
end

struct GraphTrail{T <: Int} <: AbstractGraphPath{T}
    vertices::Vector{T}
    edges::Vector{Tuple{T, T}}
    # Additional properties or methods specific to trails
end

struct GraphWalk{T <: Int} <: AbstractGraphPath{T}
    vertices::Vector{T}
    edges::Vector{Tuple{T, T}}
    # Additional properties or methods specific to walks
end
```

For each of these structures, we could add methods to construct them from a given graph, ensuring that they follow their respective constraints (e.g., no repeated edges in a trail). Here's an example **constructor** for `GraphPath`:

```julia
function GraphPath(vertices::Vector{Int}, edges::Vector{Tuple{Int, Int}})
    # Check if the number of vertices and edges are consistent.
    if length(vertices) - 1 != length(edges)
        error("Number of vertices and edges are not consistent")
    end

    # Validate that edges correctly connect the vertices.
    for (i, edge) in enumerate(edges)
        if edge != (vertices[i], vertices[i+1]) && edge != (vertices[i+1], vertices[i])
            error("Edge $edge does not connect vertices $(vertices[i]) and $(vertices[i+1])")
        end
    end

    # Additional checks like no repeated vertices or edges for a path
    if length(unique(vertices)) != length(vertices) || length(unique(edges)) != length(edges)
        error("Invalid path: repeated vertices or edges")
    end

    return GraphPath(vertices, edges)
end
```

Then we could write a new method for `==` 

```julia 
==(
        a::Union{Vector{GraphPath, Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}},
        b::Union{Vector{GraphPath, Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}},
    )::Bool
...

end 

==(
        a::Union{Vector{GraphTrail, Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}},
        b::Union{Vector{GraphTrail, Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}},
    )::Bool
...

end 

==(
        a::Union{Vector{GraphWalk, Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}},
        b::Union{Vector{GraphWalk, Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}},
    )::Bool
...

end 
```

However, we would want to be really careful if we did something like this and it **would most definitely be a separate PR if we decide to do this**.

Comment (RandyRDavila) [reply src/Invariants/Equality.jl:1]: I'll leave it up to you to decide what is best 🙂

**Comment by joshyaffee** [comment]
I thought about our options and I believe that it makes sense to override `==` on the `AbstractOptimalEdgeSet` level and I made the changes to this brach accordingly. One drawback: I assume that only undirected edge sets extend `AbstractOptimalEdgeSet`. If we want to group together both undirected and directed edge sets, I would suggest a hierarchy like this:

`AbstractOptimalEdgeSet`
 |-`AbstractOptimalUndirectedEdgeSet`
 |  |-`MinimumEdgeDominatingSet`
 |  |- ...
 |-`AbstractOptimalDirectedEdgeSet`
 |  |-...

and override `==` at the `AbstractOptimalUndirectedEdgeSet` and `AbstractOptimalDirectedEdgeSet` level for each of those abstract types. Hopefully my diagram is readable

Note that this design choice allows us to implement a separate `==` override for necessarily ordered edge sets such as walks (although this may be defined with nodes instead for simple graphs). e.g. `AbstractOptimalUndirectedEdgeSequece`. 

Finally, I see that the auto tests continue to fail. Should I be concerned?

File: test/invariant_tests.jl
Comment (RandyRDavila) [inline test/invariant_tests.jl:160]: In general try to keep the line count under 100. Maybe do this instead:

```julia
@testset "K4 Graph Tests" begin
            # any two distinct edges are valid
            edge_lst = [(1,2), (1,3), (1,4), (2,3), (2,4), (3,4)]
            true_edge_array_set = [
                [edge1, edge2] for edge1 in edge_lst 
                for edge2 in edge_lst 
                    if edge1 != edge2
            ]
            test_edge_array = compute(MinimumEdgeDominatingSet, h)
            @test any(
                [
                    test_edge_array == MinimumEdgeDominatingSet(true_edge_array) 
                    for true_edge_array in true_edge_array_set
                ]
            )
        end
        
        @testset "C5 Graph Tests" begin
            # any two edges that share no vertices are valid
            edge_lst = [
                (1,2), (1,3), (1,4), (1,5), (2,3), 
                (2,4), (2,5), (3,4), (3,5), (4,5)
            ]
            true_edge_array_set = [
                [edge1, edge2] for edge1 in edge_lst 
                for edge2 in edge_lst 
                    if edge1[1] != edge2[1] 
                        && edge1[1] != edge2[2] 
                        && edge1[2] != edge2[1] 
                        && edge1[2] != edge2[2]
            ]
            test_edge_array = compute(MinimumEdgeDominatingSet, c5)
            @test any(
                [
                    test_edge_array == MinimumEdgeDominatingSet(true_edge_array) 
                    for true_edge_array in true_edge_array_set
                ]
            )
        end

        @testset "S22 Graph Tests" begin
            true_edge_array_set = [[(1,4)]]
            test_edge_array = compute(MinimumEdgeDominatingSet, s22)
            @test any(
                [
                    test_edge_array == MinimumEdgeDominatingSet(true_edge_array) 
                    for true_edge_array in true_edge_array_set
                ]
            )
        end
```

**Review by RandyRDavila — CHANGES_REQUESTED** [review]
One more change. After you make this you can merge into `main`. I'll try and figure out why the tests are failing

File: src/Invariants/Equality.jl
Comment (RandyRDavila) [inline src/Invariants/Equality.jl:47]: The doctring needs to go here

File: src/Invariants/Equality.jl
Comment (RandyRDavila) [inline src/Invariants/Equality.jl:37]: Add this to `Invariants.jl` (see # this below):

```julia
import Base: == # this 

using Colors
using Combinatorics
using Graphs
using Graphs.SimpleGraphs: SimpleEdge # this 
```

And then remove these from `Equality.jl`. Also, change the name to `equality.jl`

**Review by RandyRDavila — CHANGES_REQUESTED** [review]
Almost missed this ....

**Comment by RandyRDavila** [comment]
@joshyaffee I changed `Equality.jl` to `equality.jl`, which fixed the tests. Super weird that this didn't show up locally when running Julia. I found this bug by looking at the failures in GitHub and then sending them to GPT4. I'm going to merge this PR now. Thanks for all the hard work!

**Pull request merged**
